feat(env): support multiline in env-file

BlackHole1 opened the pull request:
Close: https://github.com/containers/podman/issues/18724

#### Does this PR introduce a user-facing change?

```release-note
Support multiline in `--env-file`
```

/cc @Luap99 @vrothberg

BlackHole1 commented on pkg/env/env.go:
```diff
@@ -80,47 +78,108 @@ func ParseFile(path string) (_ map[string]string, err error) {
+			value = strings.ReplaceAll(value, "\\r", "\r")
+		}
+
+		// KEY*=1 => KEY, *, =1 => KEY*, =, 1
+		if separator == "*" && strings.HasPrefix(value, "=") {
+			key += "*"
+			separator = "="
+			value = strings.TrimPrefix(value, "=")
+		}
```
See: https://docs.podman.io/en/latest/markdown/podman-run.1.html#environment

<img width="682" alt="image" src="https://github.com/containers/podman/assets/8198408/ad0df4d7-0f76-4d40-a9b3-5df87e746d6c">

BlackHole1 commented on pkg/env/env.go:
```diff
@@ -80,47 +78,108 @@ func ParseFile(path string) (_ map[string]string, err error) {
+		value = strings.Trim(value, whiteSpaces)
+		value = strings.Trim(value, "\"")
+
+		if strings.HasPrefix(value, "\"") {
+			value = strings.ReplaceAll(value, "\\n", "\n")
+			value = strings.ReplaceAll(value, "\\r", "\r")
+		}
```
```diff
KEY="foo
-bar\nzzz"
+bar
+zzz"
```

BlackHole1 commented on pkg/env/env.go:
```diff
@@ -80,47 +78,108 @@ func ParseFile(path string) (_ map[string]string, err error) {
+		separator := match[2]
+		value := match[3]
+		value = strings.Trim(value, whiteSpaces)
+		value = strings.Trim(value, "\"")
```
```diff
-KEY="Value"
+KEY=Value
```

BlackHole1 commented on pkg/env/env.go:
```diff
@@ -80,47 +78,108 @@ func ParseFile(path string) (_ map[string]string, err error) {
+	re := regexp.MustCompile(
+		`(?m)(?:^|^)\s*(?:export\s+)?([\w.-]+)(\s*[=|*]\s*?|:\s+?)(\s*'(?:\\'|[^'])*'|\s*"(?:\\"|[^"])*"|\s*` +
+			"`(?:\\`|[^`])*`" + `|[^#\r\n]+)?\s*(?:#.*)?(?:$|$)`,
+	)
```
Compared to the regex used in dotenv, some modifications have been made here. `(?:\s*=\s*?|:\s+?)` has been changed to `(\s*[=|*]\s*?|:\s+?)`. This modification is made because we need to support `*` and keep the separator for subsequent logical judgments.

```diff
-(?:\s*=\s*?|:\s+?)
+(\s*[=|*]\s*?|:\s+?)
```

BlackHole1 commented:
I'm not sure why the CI is failing; it's working fine on my local machine.
<img width="1163" alt="image" src="https://github.com/containers/podman/assets/8198408/43747a8e-4266-4b51-b3a7-188b1c014de9">

CI Error:
```
[+1748s]   } PWD=/var/tmp/go/src/github.com/containers/podman/test/e2e DISABLE_HC_SYSTEMD=true CONTAINERS_REGISTRIES_CONF=/var/tmp/go/src/github.com/containers/podman/test/registries.conf GNUPGHOME= FOO=BAR]) /var/tmp/go/src/github.com/containers/podman/bin/podman --storage-opt vfs.imagestore=/tmp/imagecachedir --root /tmp/podman_test3015649375/root --runroot /tmp/podman_test3015649375/runroot --runtime runc --conmon /usr/bin/conmon --network-config-dir /etc/containers/networks --network-backend netavark --cgroup-manager systemd --tmpdir /tmp/podman_test3015649375 --events-backend file --db-backend boltdb --storage-driver vfs run --rm --env-host quay.io/libpod/alpine:latest /bin/printenv FOO
[+1748s]   Error: parsing fields in containers.conf: invalid variable: "BASH_FUNC__fmt_ctx%%=() {  local stars=\"************************************************\";\n local prefix=\"${1:-no prefix given}\";\n local message=\"${2:-no message given}\";\n echo \"$stars\";\n echo \"$prefix  ($(_ctx))\";\n echo \"$stars\"\n}"
[+1748s]   [FAILED] Failure recorded during attempt 1:
[+1748s]   Expected
[+1748s]       <int>: 125
[+1748s]   to match exit code:
[+1748s]       <int>: 0
[+1748s]   In [It] at: /var/tmp/go/src/github.com/containers/podman/test/e2e/run_env_test.go:96 @ 07/04/23 04:46:24.377
[+1748s] 
[+1748s]   Full Stack Trace
[+1748s]     github.com/containers/podman/v4/test/e2e.glob..func86.3()
[+1748s]     	/var/tmp/go/src/github.com/containers/podman/test/e2e/run_env_test.go:96 +0x2af
```

https://cirrus-ci.com/task/5001548229509120?logs=main#L3349

/cc @Luap99

Luap99 commented on pkg/env/env.go:
```diff
@@ -80,47 +78,108 @@ func ParseFile(path string) (_ map[string]string, err error) {
-	return env, scanner.Err()
+
+	// replace all \r\n and \r with \n
+	text := regexp.MustCompile(`\r\n?`).ReplaceAllString(string(content), "\n")
```
why use rexex instead of a simple string replace?

Luap99 commented on pkg/env/env.go:
```diff
@@ -80,47 +78,108 @@ func ParseFile(path string) (_ map[string]string, err error) {
+	//	KEY= => KEY, =, ""
+	//	KEY* => KEY, *, ""
+	//	KEY*=1 => KEY, *, =1
+	re := regexp.MustCompile(
```
regex should be part of the global scope as it should only be ever compiled once. And to prevent it from being compiled when it is not used use `github.com/containers/storage/pkg/regexp`

Luap99 reviewed (commented):
I am not so sure if this is not breaking to much. Keeping backwards and docker compatibility is really important. We do not know this is used today but I think the docker cli behaves the same way so changing how quotes, etc.. behave seems wrong to me.
I mainly thought this was about allowing newlines which I think can be done without breaking people.

BlackHole1 commented:
> I am not so sure if this is not breaking to much. Keeping backwards and docker compatibility is really important. We do not know this is used today but I think the docker cli behaves the same way so changing how quotes, etc.. behave seems wrong to me. I mainly thought this was about allowing newlines which I think can be done without breaking people.

Hi @Luap99. This PR is not a breaking update because the changes in this PR are backward compatible, so you can use it with confidence. You can verify this by checking the test example.

Luap99 commented:
This is what I see so far:
```
$ cat test.env 
FOO1='bar'
FOO2="bar"
FOO3=  bar
FOO4=b\nar

# current podman
$ podman run --env-file test.env alpine printenv
...
FOO3=  bar
FOO4=b\nar
FOO1='bar'
FOO2="bar"
...
# your PR
$ bin/podman run --env-file test.env alpine printenv
...
FOO3=bar
FOO4=b\nar
FOO1='bar'
FOO2=bar
...
```

BlackHole1 commented:
@Luap99 Good catch. I now need to discuss the following situations:
1. When `FOO3=  bar`, do we expect to preserve the whitespace before bar?
2. In the case of multiple lines, do we only support double quotation marks? (See below)

#### case 1

```
$ cat test.env 
FOO="aa
bb"

$ bin/podman run --env-file test.env alpine printenv FOO
aa
bb
```

#### case 2

```
$ cat test.env 
FOO="aa
bb"

$ bin/podman run --env-file test.env alpine printenv FOO
"aa
bb"
```

#### case 3

```
$ cat test.env 
FOO='aa
bb'

$ bin/podman run --env-file test.env alpine printenv FOO
aa
bb
```


#### case 4

```
$ cat test.env 
FOO=`aa
bb`

$ bin/podman run --env-file test.env alpine printenv FOO
aa
bb
```

BlackHole1 commented:
Hey @Luap99 👋. I have just completed the improvements in this PR. I fixed the issue of not preserving spaces around the value. I also addressed the problem of not removing surrounding single quotation marks in multi-line situations. Additionally, I added some additional unit tests to cover edge cases.

----

There is one final question. I have now added support for multi-line values with backticks (`), but I'm not sure if we really want to enable this.
If we don't need to support it, do we want the program to skip or throw an error when it encounters such a situation?

rhatdan commented:
LGTM

hedayat commented:
Thanks a lot for this PR. However, apparently it always replaces '\n' in a multi-line variable with a new line character (I mean replacing `\\n` with `\n`), which I suggest not to do (or do it if the value is surrounded by double quotes); which aligns with docker-compose behavior. And I guess if they are going to be replaced in double quotes case, other escape chars like `\t` should also be replaced). 

Thanks again

BlackHole1 commented:
> Thanks a lot for this PR. However, apparently it always replaces '\n' in a multi-line variable with a new line character (I mean replacing `\\n` with `\n`), which I suggest not to do (or do it if the value is surrounded by double quotes); which aligns with docker-compose behavior. And I guess if they are going to be replaced in double quotes case, other escape chars like `\t` should also be replaced).
> 
> Thanks again

Are you suggesting that deleting this line would make it consistent with Docker's behavior?
`value = strings.NewReplacer("\\n", "\n", "\\r", "\r").Replace(value)`

hedayat commented:
> Are you suggesting that deleting this line would make it consistent with Docker's behavior? `value = strings.NewReplacer("\\n", "\n", "\\r", "\r").Replace(value)`

Yes!

> 
> > BTW. I've noticed that the env-file in my local docker run doesn't support multiple lines.

It's also correct! I also find it while I was testing your branch. The interesting thing is that while docker itself doesn't seem  to properly support mult-line env files, `docker-compose` does support them and it is working fine. Actually, I was comparing with `docker-compose` behavior. (Note that I've not tested against the new `docker compose` plugin, but the older `docker-compose` one). Unloke `podman-compose`, it seems that `docker-compose` handled env files itself.

> 
> I believe it would make sense to remove `value = strings.NewReplacer("\\n", "\n", "\\r", "\r").Replace(value)`. I did this because I referred to the implementation of `dotenv`, but I realized it is unnecessary. We should not handle it this way as it can lead to unexpected bugs. Users may report a new bug: "Why did my backslash disappear?"

Yes, and it makes it impossible to embed `\n` itself inside the value (and the project I was trying to run with `podman-compose` had exactly that). And removing that line actually fixed the problem and I was able to run with `podman-compose` just fine. 

I've not tried a multi-line value surrounded by double quotes to see the `docker-compose` behavior, it is likely that it replaces escape characters in that case. But I guess it's not that important, as the user is able to use newline itself rather than using `\n` to insert a new line in the value.

rhatdan commented:
@vrothberg @Luap99 PTAL

rhatdan commented:
@vrothberg @Luap99 Reping

vrothberg commented on pkg/env/env.go:
```diff
@@ -80,47 +96,96 @@ func ParseFile(path string) (_ map[string]string, err error) {
-func parseEnv(env map[string]string, line string) error {
-	data := strings.SplitN(line, "=", 2)
+// parseEnv parse the given content into env format
+func parseEnv(env map[string]string, content string, enforceMatch bool) error {
```
Please document what `enforceMatch` does.

BlackHole1 replied:
This variable may not require additional comments to explain its purpose, as it can be inferred from its name: it throws an error if there is no match. This variable is similar to `regexp.MustCompile`.

vrothberg replied:
I wouldn't ask if it was _that_ obvious.  To understand the meaning, I had read parts of the code and look at the various callers.  It's not immediately clear.

BlackHole1 replied:
Done 😃

vrothberg replied:
Thanks!

vrothberg commented on pkg/env/env.go:
```diff
@@ -1,13 +1,29 @@
+	lineRegexp = regexp.Delayed(
+		`(?m)(?:^|^)\s*(?:export\s+)?([\w.%-]+)(\s*[=|*]\s*?|:\s+?)(\s*'(?:\\'|[^'])*'|\s*"(?:\\"|[^"])*"|\s*` +
+			"`(?:\\`|[^`])*`" + `|[^#\r\n]+)?\s*(?:#.*)?(?:$|$)`,
+	)
```
Did you consider using an external library for parsing?

I wonder whether we can use github.com/hashicorp/go-envparse and add our logic of "*" on top.  I don't know whether that will work but I would like to try it out.

The regex above is hard to maintain.

BlackHole1 replied:
TL;DR: `go-envparse` is not suitable for our needs.

I have previously looked into the `go-envparse` library, and it does not support the following scenarios:
1. It cannot handle `ENV*` prefix matching.
2. It ignores whitespace in value assignments, such as `K= V`.

Unless we fork it or patch it to support our requirements.

Currently, the regex we are using is based on https://github.com/motdotla/dotenv, with only minimal modifications.

vrothberg replied:
> Unless we fork it or patch it to support our requirements.

Forking is probably not a nice way to approach it.  In a way, we already fork parts of it (see regex) which is why I am worried.  Did you consider opening an issue at `go-envparse`?

BlackHole1 replied:
This may not be a good idea because our requirements are not 100% compatible with `dotenv`. Therefore, submitting issues or PRs to `go-envparse` is likely to be closed.

vrothberg replied:
We only know once we tried.  I think it's good practice to avoid fragmentation.  On the long run, it will be less work on the Podman as the code is maintained outside of Podman.

vrothberg replied:
@Luap99 WDYT?

Luap99 replied:
We can ask but I am not sure why they would want to support our awkward syntax. The readme clearly states:
> The following common features are intentionally missing:
>    ...

I would love to just use it as is but we must stay backwards compatible at least until 5.0. 

But I also do not really like this regex, I cannot understand it. I tried for far to long and just gave up. I know this is not your fault, regex syntax is just bad. And I know a handwritten parser is likely not any better.

vrothberg replied:
Thanks! So let's continue with this PR as is?

BlackHole1 replied:
> But I also do not really like this regex, I cannot understand it. I tried for far to long and just gave up. I know this is not your fault, regex syntax is just bad. And I know a handwritten parser is likely not any better.

I spent several hours understanding the principle of this regular expression while writing it.
Currently, we can validate whether this regular expression is backward compatible through a large number of unit tests, just as written on the side of this PR.

Luap99 replied:
Yes tests are great, as of today we have only a single basically no tests for --env-file so this is welcome.
But I really cannot comment on the regex and have no time to understand it.

I fear that we could be missing corner cases and might break existing people but well I cannot offer anything useful either so I ok with it.

What I would like is an explicit documentation on what the support syntax is in the man page. Currently all we have is `Read in a line-delimited file of environment variables.` which is not helpful to users trying to figure out how to use it.

BlackHole1 replied:
> What I would like is an explicit documentation on what the support syntax is in the man page. Currently all we have is Read in a line-delimited file of environment variables. which is not helpful to users trying to figure out how to use it.

Good idea! I think I can modify the description for `--env-file` like the following.
To prevent triggering multiple CI runs, we can check if the following description is acceptable. If there are no issues, I will submit a commit to update the content of `--env-file` description.

```md
Read the environment variables from the file, supporting prefix matching: `KEY*`, as well as multiline values in double quotes and single quotes, but not multiline values in backticks.
The env-file will ignore comments, empty lines, and spaces or tabs before and after the KEY.
If an invalid value is encountered, such as only an `=` sign, it will be skipped.
If it is a prefix match (`KEY*`), all environment variables starting with KEY on the host machine will be loaded.
If it is only KEY (`KEY`), the KEY environment variable on the host machine will be loaded.
To be compatible with Bash syntax, the `--env-file` will also support the `%%=` syntax, such as: `BASH_FUNC__echo%%=() { echo 1; }`.
```

vrothberg replied:
@BlackHole1 I don't see a man-page change in the PR yet.  Could you add that before merging?

BlackHole1 replied:
Of course, but it's the weekend now, so I will submit it on next Monday.

vrothberg replied:
Absolutely!  Enjoy the weekend :+1:

BlackHole1 replied:
@vrothberg Done 😃.

vrothberg reviewed (commented):
Nice work!  Thanks for adding tests.

TomSweeneyRedHat commented:
Can there be some documentation added to describe this new functionality?

BlackHole1 commented:
> Can there be some documentation added to describe this new functionality?

Of course, I am currently planning to modify the description of the existing `–env-file`. Please see: https://github.com/containers/podman/pull/19096#discussion_r1265403048

vrothberg commented:
@BlackHole1 are we ready to go with this PR?

BlackHole1 commented:
> @BlackHole1 are we ready to go with this PR?

Yes. Of Course.

vrothberg reviewed (approved):
LGTM
@Luap99 PTAL

Luap99 commented on docs/source/markdown/options/env-file.md:
```diff
@@ -4,4 +4,9 @@
+If an invalid value is encountered, such as only an `=` sign, it will be skipped. If it is a prefix match (`KEY*`), all environment variables starting with KEY on the host machine will be loaded.
+If it is only KEY (`KEY`), the KEY environment variable on the host machine will be loaded.
+Compatible with the `export` syntax in **dotenv**, such as: `export KEY=bar`.
+To be compatible with Bash syntax, the `--env-file` will also support the `%%=` syntax, such as: `BASH_FUNC__echo%%=() { echo 1; }`.
```
I am not sure about the last sentence? What is so special about that, I would expect any form of KEY=var to match so I don't see why this is worth a special notice.
Because it this form it is somewhat confusing what special bash syntax means.

BlackHole1 replied:
Okay. I removed it.

Luap99 reviewed (approved):
LGTM

rhatdan commented:
/hold cancel

edsantiago commented:
Heads-up everybody, this PR broke the use of octothorpe in environment variables:
```console
$ bin/podman run --env aaa='bbb#ccc' quay.io/libpod/testimage:20221018 printenv aaa
bbb    <---- should be bbb#ccc
```
A community member will be filing an issue. This is just an early notice so someone can start working on a fix.

rhatdan commented:
@BlackHole1 ^^

BlackHole1 commented:
@rhatdan @edsantiago I will be submitting a fix PR for this today.

kajinamit commented:
We noticed that we no longer able to pass a value containing new line (`\n`) and only the first line is passed to the container and that breaks some of our usage.

We are now using the CentOS Stream 9 package built from the hash 144098587706ac3fd4e257ea03c75aaf4ee29064 , while we didn't see the issue when we used the package built from hash 38e6fab9664c6e59b66e73523b307a56130316ae
, and indeed this change appears in the diffs between these two hash.

BlackHole1 commented:
> Heads-up everybody, this PR broke the use of octothorpe in environment variables:
> 
> ```
> $ bin/podman run --env aaa='bbb#ccc' quay.io/libpod/testimage:20221018 printenv aaa
> bbb    <---- should be bbb#ccc
> ```
> 
> A community member will be filing an issue. This is just an early notice so someone can start working on a fix.

Hi @edsantiago I have just fixed this issue locally. I will submit a PR in 1 hour (because I'm having lunch now 😄). This problem is quite interesting, and I will provide a detailed explanation in the PR.

BlackHole1 commented:
> We noticed that we no longer able to pass a value containing new line (`\n`) and only the first line is passed to the container and that breaks some of our usage.
> 
> We are now using the CentOS Stream 9 package built from the hash [1440985](https://github.com/containers/podman/commit/144098587706ac3fd4e257ea03c75aaf4ee29064) , while we didn't see the issue when we used the package built from hash [38e6fab](https://github.com/containers/podman/commit/38e6fab9664c6e59b66e73523b307a56130316ae) , and indeed this change appears in the diffs between these two hash.

@kajinamit Can you provide a detailed explanation of the issue you encountered? Was it related to the `--env or the --env-file?

kajinamit commented:
We are using `--env` instead of `--env-file` and the issue appears in case we pass the value containing `\n` using that option. (We are not using the podman CLI directly but calling it via ansible-podman-collections).

I've reported my issue to CentOS 9 Stream https://bugzilla.redhat.com/show_bug.cgi?id=2230212 , because I faced the issue using the podman package from CentOS 9 Stream. That contains a few more details.

BlackHole1 commented:
> We are using `--env` instead of `--env-file` and the issue appears in case we pass the value containing `\n` using that option. (We are not using the podman CLI directly but calling it via ansible-podman-collections).
> 
> I've reported my issue to CentOS 9 Stream https://bugzilla.redhat.com/show_bug.cgi?id=2230212 , because I faced the issue using the podman package from CentOS 9 Stream. That contains a few more details.

OK. I will submit the fix PR within an hour. Thank you for your feedback.
